components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -486,90 +486,360 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
+     * @return Modified attribute value
+     */
+    private String applyFilterEnhancementsForUsername(String filterOperation, String attributeName,
+            String attributeValue, ExpressionNode node) {
+
+        if (SCIMCommonUtils.isFilteringEnhancementsEnabled())
+            if (SCIMCommonConstants.EQ.equalsIgnoreCase(filterOperation))
+                if (StringUtils.equals(attributeName, SCIMConstants.UserSchemaConstants.USER_NAME_URI) && !StringUtils
+                        .contains(attributeValue, CarbonConstants.DOMAIN_SEPARATOR)) {
+                    attributeValue = UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME + CarbonConstants.DOMAIN_SEPARATOR
+                            + attributeValue;
+                    node.setValue(attributeValue);
                 }
+        return attributeValue;
+    }
 
-                if (!SCIMConstants.UserSchemaConstants.GROUP_URI.equals(attributeName)) {
-                    //get the user name of the user with this id
-                    userNames = getUserNames(attributeName, filterOperation, attributeValue);
-                } else {
-                    if (filterOperation.equalsIgnoreCase(SCIMCommonConstants.EQ)) {
-                        userNames = carbonUM.getUserListOfRole(attributeValue);
-                    } else if (carbonUM instanceof AbstractUserStoreManager) {
-                        String[] roleNames = getRoleNames(filterOperation, attributeValue);
-                        userNames = getUserListOfRoles(roleNames);
-                    } else {
-                        throw new NotImplementedException(String.format("Filter operator %s is not supported " +
-                                "by the user store.", filterOperation));
-                    }
+    /**
+     * Update the domain parameter from the domain in attribute value and update the Value in Expression node
+     *
+     * @param domainName Domain name in the filter request
+     * @param node       Expression node
+     * @return Domain name extracted from the attribute value
+     * @throws BadRequestException Domain miss match in domain parameter and attribute value
+     */
+    private String resolveDomainInAttributeValueAndUpdateExpressionNode(String domainName, ExpressionNode node)
+            throws BadRequestException {
+
+        String extractedDomain;
+        String attributeName = node.getAttributeValue();
+        String filterOperation = node.getOperation();
+        String attributeValue = node.getValue();
+
+        if (isDomainNameEmbeddedInAttributeValue(filterOperation, attributeName, attributeValue)) {
+            int indexOfDomainSeparator = attributeValue.indexOf(CarbonConstants.DOMAIN_SEPARATOR);
+            extractedDomain = attributeValue.substring(0, indexOfDomainSeparator).toUpperCase();
+
+            // Update then newly extracted attribute value in the expression node.
+            node.setValue(attributeValue.substring(indexOfDomainSeparator + 1));
```

Comment:
It is more readable if we assign attributeValue.substring(indexOfDomainSeparator + 1 to a variable and use it here

Reply:
fix - ac4cf58d8f20ee81c08cb55ec47eed9ce3c305ed